python/whylogs/core/metrics/multimetric.py

```diff
@@ -0,0 +1,170 @@
+    processing of the logged values and track serveral metrics on
+    the results. The sub-metrics must either be a StandardMetric, or tagged
+    as a @custom_metric or @dataclass. Note that MultiMetric is neither, so it
+    cannot be nested.
+
+    Typically you will need to override namespace(); columnar_update(), calling
+    it on the submetrics as needed; and the zero() method to return an
+    appropriate "empty" instance of your metric. You will need to override from_protobuf()
+    and merge() if your subclass __init__() method takes arguments different than
+    MultiMetrtic's. You can use the submetrics_from_protbuf() and merge_submetrics()
+    helper methods to implement them. The MultiMetric class will handle the rest of
+    the Metric interface. Don't use / or : in the subclass' namespace.
+
+    See UnicodeRangeMetric for an example.
+    """
+
+    submetrics: Dict[str, Dict[str, Metric]]
+
+    def __init__(self, submetrics: Dict[str, Dict[str, Metric]]):
+        """
+                Parameters
+                ----------
+                submetrics : Dict[str, Dict[str, Metric]]
+                    The collection of metrics that comprise the CompoundMetric.
+
+        TODO: type this
+
+                       "<submetric name>/<component name>"
+                    Submetric names should only contain alphanumeric characters,
+                    hyphens, and underscores.
+        """
+
+        if ":" in self.namespace or "/" in self.namespace:
+            raise ValueError(f"Invalid namespace {self.namespace}")
+        for sub_name in submetrics.keys():
+            if ":" in sub_name or "/" in sub_name:
+                raise ValueError(f"Invalid submetric name {sub_name}")
+
+        self.submetrics = deepcopy(submetrics)
+
+    def merge_submetrics(self: MULTI_METRIC, other: MULTI_METRIC) -> Dict[str, Metric]:
+        if self.namespace != other.namespace:
+            raise ValueError(f"Attempt to merge CompoundMetrics {self.namespace} and {other.namespace}")
+
+        submetric_names = set(self.submetrics.keys())
+        submetric_names.update(other.submetrics.keys())
+        submetrics: Dict[str, Dict[str, Metric]] = dict()
+        for submetric_name in submetric_names:
+            if submetric_name in self.submetrics and submetric_name in other.submetrics:
+                submetrics[submetric_name] = _do_submetric_merge(
+                    self.submetrics[submetric_name], other.submetrics[submetric_name]
+                )
+            elif submetric_name in self.submetrics:
+                submetrics[submetric_name] = deepcopy(self.submetrics[submetric_name])
+            else:
+                submetrics[submetric_name] = deepcopy(other.submetrics[submetric_name])
+
+        return submetrics
+
+    def merge(self: MULTI_METRIC, other: MULTI_METRIC) -> MULTI_METRIC:
+        return self.__class__(self.merge_submetrics(other))
+
+    def to_protobuf(self) -> MetricMessage:
+        msg = {}
```

Comment:
is it a valid case for submetrics to be empty?

Reply:
Hmmm... it works for most methods. How does `MetricMessage` feel about being empty?

Reply:
I was wondering if the code here handles the empty case ok, jut curious about the edge case.